### PR TITLE
feat: autoscaling changes to support multi partition

### PR DESCRIPTION
**Pull request by yhl25**
- cache daemon clients to avoid creating new clients each time. (LRU based)
- supports autoscaling when there are multiple partitions by considering processing rates and pending information of all the partitions

### CODE REVIEW

File: pkg/reconciler/vertex/scaling/client_cache.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+	daemonclient "github.com/numaproj/numaflow/pkg/daemon/client"
+)
+
+type entry struct {
+	address string
+	client  *daemonclient.DaemonClient
+}
+
+// ClientCache is a cache for daemon clients.
```
Comment (vigith) [inline pkg/reconciler/vertex/scaling/client_cache.go:15]: please comment what kind of cache you have implemented. also.. can we make it generic?

File: pkg/reconciler/vertex/scaling/scaling.go [hunk 2]
```diff
@@ -61,6 +62,8 @@ func NewScaler(client client.Client, opts ...Option) *Scaler {
 	}
 	vertexMetricsCache, _ := lru.New(10000)
 	s.vertexMetricsCache = vertexMetricsCache
+	// cache top 100 daemon clients
+	s.daemonClientsCache = NewClientCache(100)
```
Comment (vigith) [inline pkg/reconciler/vertex/scaling/scaling.go:66]: make 100 optional

File: pkg/reconciler/vertex/scaling/client_cache.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	}
+	return nil
+}
+
+// DeleteAll function closes all the daemon clients in the cache and removes them from the cache.
+func (c *ClientCache) DeleteAll() {
+	c.Lock()
+	defer c.Unlock()
+	for _, element := range c.clientsMap {
+		_ = element.Value.(*entry).client.Close()
```
Comment (vigith) [inline pkg/reconciler/vertex/scaling/client_cache.go:61]: this cannot be generic as is.. so, pass a func that can clean it up.

File: pkg/reconciler/vertex/scaling/client_cache.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+	daemonclient "github.com/numaproj/numaflow/pkg/daemon/client"
+)
+
+type entry struct {
+	address string
+	client  *daemonclient.DaemonClient
+}
+
+// ClientCache is a cache for daemon clients.
+type ClientCache struct {
```
Comment (whynowy) [inline pkg/reconciler/vertex/scaling/client_cache.go:16]: Why don't use a simple LRU cache?

Comment (vigith) [reply pkg/reconciler/vertex/scaling/client_cache.go:16]: I think it is because he has that DeleteAll func that closes the client.

Comment (whynowy) [reply pkg/reconciler/vertex/scaling/client_cache.go:16]: There's a `NewWithEvict` function which can give an eviction callback, the callback is also called in `Purge()`.

```golang
// NewWithEvict constructs a fixed size cache with the given eviction
// callback.
func NewWithEvict(size int, onEvicted func(key interface{}, value interface{})) (*Cache, error) {
	lru, err := simplelru.NewLRU(size, simplelru.EvictCallback(onEvicted))
	if err != nil {
		return nil, err
	}
	c := &Cache{
		lru: lru,
	}
	return c, nil
}
```

Comment (yhl25) [reply pkg/reconciler/vertex/scaling/client_cache.go]: didn't know about this.. updated it to use simple LRU.

**Review by vigith — APPROVED** [review]

File: pkg/reconciler/vertex/scaling/scaling.go [hunk 5]
```diff
@@ -202,50 +208,66 @@ func (s *Scaler) scaleOneVertex(ctx context.Context, key string, worker int) err
...
-	if !existing || rate < 0 { // Rate not available
-		log.Debugf("Vertex %s has no rate information, skip scaling", vertex.Name)
-		return nil
-	}
-	pending, existing := vMetrics[0].Pendings["default"]
-	if !existing || pending < 0 || pending == isb.PendingNotAvailable {
-		// Pending not available, we don't do anything
-		log.Debugf("Vertex %s has no pending messages information, skip scaling", vertex.Name)
-		return nil
+	// vmMetrics is a map which contains metrics of all the partitions of a vertex.
```
Comment (whynowy) [inline pkg/reconciler/vertex/scaling/scaling.go:227]: Minor: `vMetrics`.

**Review by whynowy — APPROVED** [review]

**Pull request merged**